Implement Variables in the DSL

jacknugent1529 opened the pull request:
Variables are implemented as a pre-parsing addition to the language, so they are evaluated before the language itself is parsed. This means that variables will not conflict with the grammar of the language. The exact specifications are described in issue #998, but the basic features are the following
- multi-line variables declared with triple quotes (e.g. `$var = """<content>"""`
- escaped-string variables declared with single quotes (e.g. `$var = "This will evaluate to a string surrounded by quotes"`)
- Support for template strings similar to python f-strings

The main files added in this pull request are `vars_parser.py` and `vars.lark` which evaluate the  variables in the dsl file. Additionally, there is an example file using variables in several different manners, and the `vars_test.py` file for testing purposes. This pull request will close issue #998.

john-hahn reviewed (commented):
The changes made to the files all pertain to the goal of implementing variables in the DSL. First off, the `src/dsl/src/vars.lark` file which sets the rules for the variables, nicely and efficiently denotes what is recognized as triple quotes, escaped strings, key values. `src/dsl/src/vars_parser.py` takes the variables and manipulates it using the functions defined. They essentially make the tree more readable, and Jack has done a nice job with the coding of these functions. Lastly, there is a file for testing the variables out. The `src/dsl/src/dsl_parser.py` file has minimal change, which is the inclusion of the `evalVars` library.

Overall, all these changes are necessary for the implementation of variables, and there isn't any problem I see with the code.

ecoble commented on src/dsl/examples/vars.dsl:
```diff
@@ -0,0 +1,25 @@
+
+
+GAME START Room A END room C
+  intro: $intro {firstname: John, lastname: "Smith"}
```
Is it intentional that Smith is in quotations and John is not? If so, it might be good to leave a note pointing out that you are showing the flexibility of DSL

jacknugent1529 replied:
Yes, it is intentional. The quotes are only necessary if escape characters are included. I included a comment to that effect, and it will be mentioned in the documentation.

ecoble commented on src/dsl/src/vars.lark:
```diff
@@ -0,0 +1,40 @@
+// ?start: assignment
```
This comment looks like it should be removed, or turned into prose

jacknugent1529 replied:
Yeah, it was not necessary, so I removed it

ecoble commented on src/dsl/src/vars_test.py:
```diff
@@ -0,0 +1,17 @@
+# simple script to test the functionality of the evalVars function
```
This and any other tests file should go in their own tests directory, for now I think you can just nest the tests directory within your src directory, I will check about whether you should put it in the main tests directory since you are using python.

jacknugent1529 replied:
This script only printed out the file after the variables had been replaced by the appropriate strings. Since it wasn't a true test file, I instead included the functionality in the `dsl_parser.py` file with the `--debug` flag.

ecoble reviewed (changes requested):
This looks great! Just a few small things.

jacknugent1529 commented:
Closing statement: We added support for variables in the dsl. This was implemented as a pre-parsing transformation that replaces references to variables with the variables itself. The main features involve escaped string variables, multi-line variables, and template strings similar to python f-strings. Since opening the pull request, we simplified debugging functionality to use the `--debug` flag with the `dsl_parser.py` instead of needing another script. Since these changes have been successfully merged, this closes issue #998.

ecoble commented:
Issue Score: ✔️++

Comments:
Great work adding variables to the DSL, and documenting progress on this PR!

<!-- {"score": "check-plus-plus"} -->
